ignite/metrics/psnr.py

```diff
@@ -0,0 +1,118 @@
+    The output of the engine's ``process_function`` needs to be in format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        psnr = PSNR()
+        psnr.attach(engine, "psnr")
+        # ...
+        state = engine.run(data)
+        print(f"PSNR: {state.metrics['psnr']}")
+
+    .. versionadded:: 0.5.0
+    """
+
+    def __init__(
+        self,
+        data_range: Optional[Union[int, float]] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        super().__init__(output_transform=output_transform, device=device)
+        self.data_range = data_range
+
+    def _check_shape_dtype(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred, y = output
+        if y_pred.dtype != y.dtype:
+            raise TypeError(
+                f"Expected y_pred and y to have the same data type. Got y_pred: {y_pred.dtype} and y: {y.dtype}."
+            )
+
+        if y_pred.shape != y.shape:
+            raise ValueError(
+                f"Expected y_pred and y to have the same shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
+            )
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_batchwise_psnr = 0  # type: Union[int, torch.Tensor]
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        self._check_shape_dtype(output)
+        y_pred, y = output[0].detach(), output[1].detach()
+        data_range = self.data_range
+
+        if data_range is None:
+            dmin, dmax = _dtype_range[y.dtype]
+            true_min, true_max = y.min(), y.max()
+            if true_max > dmax or true_min < dmin:
+                raise ValueError(
+                    "y has intensity values outside the range expected "
+                    "for its data type. Please manually specify the `data_range`."
+                )
+            if true_min >= 0:
+                # most common case (255 for uint8, 1 for float)
+                data_range = dmax
+            else:
+                data_range = dmax - dmin
+
+        mse_error = torch.pow(y_pred.double() - y.view_as(y_pred).double(), 2).mean()
+        self._sum_of_batchwise_psnr += 10.0 * torch.log10(data_range ** 2 / mse_error).to(device=self._device)
+        self._num_examples += y.shape[0]
+
+    @sync_all_reduce("_sum_of_batchwise_psnr", "_num_examples")
+    def compute(self) -> Union[float, torch.Tensor]:
+        if self._num_examples == 0:
+            raise NotComputableError("PSNR must have at least one example before it can be computed.")
+        return self._sum_of_batchwise_psnr / self._num_examples
+
+
+_int_types = (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64)
```

Comment:
I think we should either support only uint8 and floats as dtype or do as for SSIM where data_range is required parameter.

For example, I'm trying that:
```python
torch.manual_seed(1)
y = torch.randint(0, 256, size=(32, 3, 32, 32))
y_pred = torch.randint(0, 256, size=(32, 3, 32, 32))

p.reset()
p.update((y_pred, y))
p.compute()
```
and see the error: `RuntimeError: Overflow when unpacking long`. As error message is not indicating the source of the error, it is almost impossible to understand what happens.

Reply:
So, shall we hardcode with `(-1, 1)` and `(0, 255)`?